### ISSUE TITLE
Bug: ESLint raises type-safety warnings for imported functions and types from shared libraries

### ISSUE THREAD

**Issue by lucas-barake** [typescript-eslint/typescript-eslint#8855]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Issue Description

I recently encountered an issue where ESLint raises type-safety related warnings when importing and using functions or types from shared libraries, even when the usage is correct according to the library's type definitions.

I initially reported this issue to the Nx repository ([https://github.com/nrwl/nx/issues/22620](https://github.com/nrwl/nx/issues/22620)), but after further investigation, it seems that the issue stems from the @typescript-eslint plugin.

Here's a minimal reproduction of the issue:

1. I have a shared library (`libs/utils`) with the following code:

```ts
import { create } from 'zustand';

type ZustandStore = {
  someValue: string;
}
export const useZustandStore = create<ZustandStore>(() => ({
  someValue: 'Hello, Zustand!',
}));
```

2. In my application (`apps/web`), I import and use the `useZustandStore` hook:

```tsx
import React from 'react';
import { useZustandStore } from '@myorg/utils';

export const App: React.FC = () => {
  // ESLint: Unsafe member access .someValue on an `any` value.(@typescript-eslint/no-unsafe-member-access)
  // ESLint: Unsafe call of an `any` typed value.(@typescript-eslint/no-unsafe-call)
  // ESLint: Unsafe assignment of an `any` value.(@typescript-eslint/no-unsafe-assignment)
  const [message, setMessage] = React.useState<string>("Hello, Nx!");

  // ESLint: Unsafe return of an `any` typed value.(@typescript-eslint/no-unsafe-return)
  // ESLint: Unsafe member access .someValue on an `any` value.(@typescript-eslint/no-unsafe-member-access)
  // ESLint: Unsafe call of an `any` typed value.(@typescript-eslint/no-unsafe-call)
  const store = useZustandStore((state) => state.someValue);

  return (
    <div>
      <h1>{message}</h1>
    </div>
  );
}
```

These warnings occur even though the usage of the Zustand store is correct according to the library's type definitions.

The Nx team mentioned that the `getConstrainedTypeAtLocation` function used by these rules is returning any when parsing the imported functions and types, which seems to be the root cause of the issue.

Thank you for your time and assistance!

### Reproduction Repository Link

https://github.com/lucas-barake/nx-eslint-bug

### Repro Steps

1) clone the repo
2) `yarn install`
3) `yarn run lint`

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `^7.3.0` |
| `@typescript-eslint/parser`            | `^7.3.0` |
| `TypeScript`                           | `~5.4.2` |
| `ESLint`                               | `~8.57.0` |
| `node`                                 | `20.11.1` |

**Comment by bradzacher** [comment]
Are you able to reduce the reproduction at all?
At a quick glance currently there are dozens of files that are irrelevant and the config contains many plugins and rules that are unrelated.

We maintainers are bandwidth constrained so the bigger the reproduction the harder it is for us to help.

**Comment by lucas-barake** [comment]
Hey @bradzacher, thanks for getting back to me so quickly!

I've gone ahead and trimmed down the ESLint rules to just the relevant ones. As for the project itself, since it's an Nx monorepo and the issue happens with the shared libraries, there's not a whole lot I can do to simplify it further. The key parts are the root `.eslintrc.json` and the one it extends at `apps/myorg/.eslintrc.json`.

To provide a bit more context, [`apps/myorg/src/app/app.tsx`](https://github.com/lucas-barake/nx-eslint-bug/blob/main/apps/myorg/src/app/app.tsx) consumes the shared library that contains the example Zustand stores at [`libs/utils/src/lib/utils.ts`](https://github.com/lucas-barake/nx-eslint-bug/blob/main/libs/utils/src/lib/utils.ts).

I believe I've got the reproduction as concise as possible while still demonstrating the issue. If you have any other questions or need any additional clarification, please don't hesitate to let me know.

Thanks again for your help - I'm really looking forward to your insights on this!

**Comment by auvred** [comment]
> I believe I've got the reproduction as concise as possible while still demonstrating the issue. If you have any other questions or need any additional clarification, please don't hesitate to let me know.

There are still prettier, nx, vite, vitest and several eslint plugins unrelated to the typescript-eslint

---

At first glance, there's a lot of tsconfigs here, could you try to enable `EXPERIMENTAL_useProjectService` in the root `.eslintrc.json`?

https://typescript-eslint.io/packages/parser/#experimental_useprojectservice

```diff
      "parserOptions": {
+       "EXPERIMENTAL_useProjectService": true
-       "project": "./tsconfig.base.json"
      },
```

Maybe this will solve the problem (if it works, this will show that parserSettings is not configured correctly).

**Comment by lucas-barake** [comment]
Yep, that fixed it.

Thanks!
